component/azstorage/block_blob.go

```diff
@@ -120,97 +120,60 @@ func (bb *BlockBlob) UpdateConfig(cfg AzStorageConfig) error {
+	bb.Auth = getAzAuthT2(bb.Config.authConfig)
 	if bb.Auth == nil {
-		log.Err("BlockBlob::getCredential : Failed to retrieve auth object")
-		return nil
+		log.Err("BlockBlob::getServiceClient : Failed to retrieve auth object")
+		return nil, fmt.Errorf("failed to retrieve auth object")
 	}
 
-	cred := bb.Auth.getCredential()
-	if cred == nil {
-		log.Err("BlockBlob::getCredential : Failed to get credential")
-		return nil
+	svcClient, err := bb.Auth.getServiceClient(&bb.Config)
+	if err != nil {
+		log.Err("BlockBlob::getServiceClient : Failed to get service client [%s]", err.Error())
+		return nil, err
 	}
 
-	return cred.(azblob.Credential)
-}
-
-// NewPipeline creates a Pipeline using the specified credentials and options.
-func NewBlobPipeline(c azblob.Credential, o azblob.PipelineOptions, ro ste.XferRetryOptions) pipeline.Pipeline {
-	// Closest to API goes first; closest to the wire goes last
-	f := []pipeline.Factory{
-		azblob.NewTelemetryPolicyFactory(o.Telemetry),
-		azblob.NewUniqueRequestIDPolicyFactory(),
-		ste.NewBlobXferRetryPolicyFactory(ro),
-	}
-	f = append(f, c)
-	f = append(f,
-		pipeline.MethodFactoryMarker(), // indicates at what stage in the pipeline the method factory is invoked
-		ste.NewRequestLogPolicyFactory(ste.RequestLogOptions{
-			LogWarningIfTryOverThreshold: o.RequestLog.LogWarningIfTryOverThreshold,
-			SyslogDisabled:               o.RequestLog.SyslogDisabled,
-		}))
-
-	return pipeline.NewPipeline(f, pipeline.Options{HTTPSender: o.HTTPSender, Log: o.Log})
+	return svcClient.(*service.Client), nil
 }
 
 // SetupPipeline : Based on the config setup the ***URLs
 func (bb *BlockBlob) SetupPipeline() error {
 	log.Trace("BlockBlob::SetupPipeline : Setting up")
 	var err error
 
-	// Get the credential
-	cred := bb.getCredential()
-	if cred == nil {
-		log.Err("BlockBlob::SetupPipeline : Failed to get credential")
-		return errors.New("failed to get credential")
-	}
-
-	// Create a new pipeline
-	options, retryOptions := getAzBlobPipelineOptions(bb.Config)
-	bb.Pipeline = NewBlobPipeline(cred, options, retryOptions)
-	if bb.Pipeline == nil {
-		log.Err("BlockBlob::SetupPipeline : Failed to create pipeline object")
-		return errors.New("failed to create pipeline object")
-	}
-
-	// Get the endpoint url from the credential
-	bb.Endpoint, err = url.Parse(bb.Auth.getEndpoint())
+	// create the service client
+	bb.Service, err = bb.getServiceClient()
 	if err != nil {
-		log.Err("BlockBlob::SetupPipeline : Failed to form base end point url [%s]", err.Error())
-		return errors.New("failed to form base end point url")
+		log.Err("BlockBlob::SetupPipeline : Failed to get service client [%s]", err.Error())
+		return err
 	}
 
-	// Create the service url
-	bb.Service = azblob.NewServiceURL(*bb.Endpoint, bb.Pipeline)
-
-	// Create the container url
-	bb.Container = bb.Service.NewContainerURL(bb.Config.container)
-
+	// create the container client
+	bb.Container = bb.Service.NewContainerClient(bb.Config.container)
```

Comment:
Do we need all three methods one creating service client, one creating container client, and this one which also creates container client. or we can create some common method and use it as needed.

Reply:
We are only creating service client here. Container client is created from service client so that it uses the same pipeline options.
`bb.Service.NewContainerClient(bb.Config.container)` call goes to SDK.

Reply:
There are three methods in a row which are creating clients so just see if we can extract few things out into a seperate method.

Reply:
The three methods are,
`NewServiceClient`: this method is for dynamically updating the service client when SAS in the config is updated.
`getServiceClient`: this is the method where we set up the `Auth` and create the service client
`SetupPipeline`: this method calls the getServiceClient() to create the service client and then the container client from it.

Reply:
I can rename the methods if that makes sense. Like,
`NewServiceClient` to `UpdateServiceClient`
`getServiceClient` to `createServiceClient`